### ISSUE TITLE
No definition found for type 'androidx.lifecycle.SavedStateHandle'

### ISSUE THREAD

**Issue by juan-wesper** [InsertKoinIO/koin#2044]
**Describe the bug**
This used to work fine for v3.5.6, but we've seen this after upgrading to v4.0.0.

**To Reproduce**
1. Have a ViewModel that takes `SavedStateHandle` as param.
2. Use `viewModelOf` (the one from the new package) on your module.
3. Turn 'Don't keep activities' from the Developer Options on your phone settings.
4. Open the screen in which the aforementioned ViewModel is used.

Note that this does not always happen

**Expected behavior**
The ViewModel is instantiated correctly

**Actual behavior**
The app crashes with the following exception:
```
org.koin.core.error.InstanceCreationException: Could not create instance for '[Factory: '<redacted>.ui.screens.test.upload.UploadViewModel']'
at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:58)
at org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:38)
at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:110)
at org.koin.core.scope.Scope.resolveFromRegistry(Scope.kt:321)
at org.koin.core.scope.Scope.resolveFromContext(Scope.kt:311)
at org.koin.core.scope.Scope.stackParametersCall(Scope.kt:281)
at org.koin.core.scope.Scope.resolveInstance(Scope.kt:259)
at org.koin.core.scope.Scope.resolveWithOptionalLogging(Scope.kt:232)
at org.koin.core.scope.Scope.get(Scope.kt:215)
at org.koin.viewmodel.factory.KoinViewModelFactory.create(KoinViewModelFactory.kt:39)
at androidx.lifecycle.viewmodel.ViewModelProviderImpl_androidKt.createViewModel(ViewModelProviderImpl.android.kt:34)
at androidx.lifecycle.viewmodel.ViewModelProviderImpl.getViewModel$lifecycle_viewmodel_release(ViewModelProviderImpl.kt:65)
at androidx.lifecycle.viewmodel.ViewModelProviderImpl.getViewModel$lifecycle_viewmodel_release$default(ViewModelProviderImpl.kt:47)
at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.android.kt:91)
at org.koin.viewmodel.GetViewModelKt.resolveViewModel(GetViewModel.kt:60)
at <redacted>.ui.screens.test.upload.UploadScreenKt.UploadScreen(UploadScreen.kt:120)
at <redacted>.ui.navigation.ComposableSingletons$NavGraphKt$lambda-20$1.invoke(NavGraph.kt:157)
at <redacted>.ui.navigation.ComposableSingletons$NavGraphKt$lambda-20$1.invoke(NavGraph.kt:154)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:139)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
at androidx.navigation.compose.NavHostKt$NavHost$32$1.invoke(NavHost.kt:694)
at androidx.navigation.compose.NavHostKt$NavHost$32$1.invoke(NavHost.kt:693)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:109)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
at androidx.compose.runtime.CompositionLocalKt.CompositionLocalProvider(CompositionLocal.kt:401)
at androidx.compose.runtime.saveable.SaveableStateHolderImpl.SaveableStateProvider(SaveableStateHolder.kt:85)
at androidx.navigation.compose.NavBackStackEntryProviderKt.SaveableStateProvider(NavBackStackEntryProvider.kt:65)
at androidx.navigation.compose.NavBackStackEntryProviderKt.access$SaveableStateProvider(NavBackStackEntryProvider.kt:1)
at androidx.navigation.compose.NavBackStackEntryProviderKt$LocalOwnersProvider$1.invoke(NavBackStackEntryProvider.kt:52)
at androidx.navigation.compose.NavBackStackEntryProviderKt$LocalOwnersProvider$1.invoke(NavBackStackEntryProvider.kt:51)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:109)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
at androidx.compose.runtime.CompositionLocalKt.CompositionLocalProvider(CompositionLocal.kt:380)
at androidx.navigation.compose.NavBackStackEntryProviderKt.LocalOwnersProvider(NavBackStackEntryProvider.kt:47)
at androidx.navigation.compose.NavHostKt$NavHost$32.invoke(NavHost.kt:693)
at androidx.navigation.compose.NavHostKt$NavHost$32.invoke(NavHost.kt:674)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:139)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1$5.invoke(AnimatedContent.kt:803)
at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1$5.invoke(AnimatedContent.kt:792)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:118)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
at androidx.compose.animation.AnimatedVisibilityKt.AnimatedEnterExitImpl(AnimatedVisibility.kt:771)
at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1.invoke(AnimatedContent.kt:774)
at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1.invoke(AnimatedContent.kt:757)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:109)
at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
at androidx.compose.animation.AnimatedContentKt.AnimatedContent(AnimatedContent.kt:816)
at androidx.navigation.compose.NavHostKt.NavHost(NavHost.kt:646)
at androidx.navigation.compose.NavHostKt$NavHost$34.invoke(Unknown Source:29)
at androidx.navigation.compose.NavHostKt$NavHost$34.invoke(Unknown Source:10)
at androidx.compose.runtime.RecomposeScopeImpl.compose(RecomposeScopeImpl.kt:192)
at androidx.compose.runtime.ComposerImpl.recomposeToGroupEnd(Composer.kt:2825)
at androidx.compose.runtime.ComposerImpl.skipCurrentGroup(Composer.kt:3116)
at androidx.compose.runtime.ComposerImpl.doCompose(Composer.kt:3607)
at androidx.compose.runtime.ComposerImpl.recompose$runtime_release(Composer.kt:3552)
at androidx.compose.runtime.CompositionImpl.recompose(Composition.kt:948)
at androidx.compose.runtime.Recomposer.performRecompose(Recomposer.kt:1206)
at androidx.compose.runtime.Recomposer.access$performRecompose(Recomposer.kt:132)
at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:616)
at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:585)
at androidx.compose.ui.platform.AndroidUiFrameClock$withFrameNanos$2$callback$1.doFrame(AndroidUiFrameClock.android.kt:41)
at androidx.compose.ui.platform.AndroidUiDispatcher.performFrameDispatch(AndroidUiDispatcher.android.kt:109)
at androidx.compose.ui.platform.AndroidUiDispatcher.access$performFrameDispatch(AndroidUiDispatcher.android.kt:41)
at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.doFrame(AndroidUiDispatcher.android.kt:69)
at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1299)
at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1309)
at android.view.Choreographer.doCallbacks(Choreographer.java:923)
at android.view.Choreographer.doFrame(Choreographer.java:847)
at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1283)
at android.os.Handler.handleCallback(Handler.java:942)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loopOnce(Looper.java:226)
at android.os.Looper.loop(Looper.java:313)
at android.app.ActivityThread.main(ActivityThread.java:8757)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [androidx.compose.runtime.PausableMonotonicFrameClock@2593ecf, androidx.compose.ui.platform.MotionDurationScaleImpl@43d005c, StandaloneCoroutine{Cancelling}@73d5f65, AndroidUiDispatcher@550943a]
Caused by: org.koin.core.error.NoDefinitionFoundException: No definition found for type 'androidx.lifecycle.SavedStateHandle'. Check your Modules configuration and add missing type and/or qualifier!
at org.koin.core.scope.Scope.resolveFromContext(Scope.kt:532)
at org.koin.core.scope.Scope.stackParametersCall(Scope.kt:273)
at org.koin.core.scope.Scope.resolveInstance(Scope.kt:259)
at org.koin.core.scope.Scope.resolveWithOptionalLogging(Scope.kt:232)
at org.koin.core.scope.Scope.get(Scope.kt:215)
at <redacted>.ui.UiModuleKt$uiModule$lambda$33$$inlined$viewModelOf$default$27.invoke(ViewModelOf.kt:226)
at <redacted>.ui.UiModuleKt$uiModule$lambda$33$$inlined$viewModelOf$default$27.invoke(ViewModelOf.kt:99)
at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:51)
```

**Koin module and version:**
io.insert-koin:koin-androidx-compose:4.0.0

**Snippet or Sample project to help reproduce**
Add a snippet or even a small sample project to hel reproduce your case.

**Comment by marcardar** [comment]
Seeing something similar in crashlytics (S22 Ultra, Android 14) though unable to reproduce locally (even after trying Don't Keep Activities). Instead, I'm using `koinViewModel()` with `Activity` `viewModelStoreOwner`.

**Comment by marcardar** [comment]
Is it related? https://github.com/InsertKoinIO/koin/issues/1009

**Comment by florent37** [comment]
Crashlytics contains a lot of issues related to koin / viewmodels / SavedStateHandle also, it's not reproducible locally but it's on top of our crashs 

<img width="937" alt="Capture d’écran 2024-12-04 à 08 34 17" src="https://github.com/user-attachments/assets/510017aa-ec53-49b6-bef8-2f930ddd4d15">

**Comment by arnaudgiuliani** [comment]
Ok. I"m on it.
@florent37 is i same stacktraces as #1625 ?

**Comment by arnaudgiuliani** [comment]
@juan-wesper any particular pro-guard/minification stuff?

I tried on NowInAndroid with Don't Keep Activities, with minified release build. Couldn't reproduce for now.
SSH seems to not be triggered :/

**Comment by arnaudgiuliani** [comment]
I see. the use of `<redacted>` plugin, right? It could affect class research, especially on ViewModel and SavedStateHandle injection.

@florent37 do you use any extra plugin like redacted?

**Comment by florent37** [comment]
@arnaudgiuliani nope we are not using it

**Comment by marcardar** [comment]
I've experienced the bug both with and without obfuscating `androidx.lifecycle.*`

**Comment by arnaudgiuliani** [comment]
@marcardar do you have a snippet to reproduce it?

**Comment by marcardar** [comment]
The viewmodel is like this:
```
class AppScreenModel(
    // placed in first position because this workaround was mentioned in some similar issue
    private val savedStateHandle: SavedStateHandle, 
    ...,
) : ViewModel() {
    ...,
}

@Composable
fun AppScreen(
    appScreenModel: AppScreenModel = koinActivityViewModel(),
    ...
) {
    ...
}

@Composable
fun AppNavHost(...) {
    NavHost(...) {
        composable(route = "resultsScreen") {
            val appScreenModel: AppScreenModel = koinActivityViewModel()
            ...
        }
    }
}
```
So actually it is called in two different places. Could that be an issue?

Note, this happens very rarely in my app, and I haven't been able to reproduce.

The other thing is that my app has two variants with very similar codebase (identical, regarding AppScreen and AppScreenModel) but the error only happen in one variant. So it makes me wonder if it's some dex thing?

**Comment by arnaudgiuliani** [comment]
I'v added a fix to make SavedStateHandle injection via direct parameter calls instead of lambda: ae940219feba4f9042e7a965775dd928dc3552f2

Let's see how it goes, if some of you can embed next build?

**Comment by arnaudgiuliani** [comment]
See also https://github.com/InsertKoinIO/koin/issues/1935
about injecting SavedStateHandle & Nav backstack entries

**Comment by arnaudgiuliani** [comment]
Ok, let's see that for 4.0.3

**Comment by arnaudgiuliani** [comment]
@AmrKhalilDH I need more details config & what code is used. Can be different causes.

**Comment by arnaudgiuliani** [comment]
Let's close this issue, and reopen in 4.1.0 if needed. If you are using navigation, take care about your navController to not leak it. 
Activate Koin debug logs, and look at such patterns. It indicates that you are resolving a SavedStateHandle.

<img width="649" alt="Image" src="https://github.com/user-attachments/assets/129d276b-90b2-42b2-a78a-ffabc425e938" />

**Comment by okycelt** [comment]
For us, this happens with 4.0.2 and 4.0.4. We didn't get any crash reports from 3.5.0.

We can reproduce it "reliably", but only in one of our apps. It always happens on the first start of a release build with the "root" view model. It doesn't matter whether minification is enabled. It's a single-activity (Android-only) compose app looking like this (koin is started in `Application.onCreate()`).

```kotlin
class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContent {
            KoinAndroidContext {
            	AppContent()
            }
        }
    }
}

@Composable
fun AppContent(viewModel: RootViewModel = koinViewModel()) {
    // ...
}

class RootViewModel(
    // 11 more dependencies
    private val savedStateHandle: SavedStateHandle,
    private val defaultDispatcher: CoroutineDispatcher = Dispatchers.Default
) : ViewModel() 
```

It seems like something that is supposed to inject `SavedStateHandle` isn't initialized yet when `koinViewModel()` is called, because if we inject the view model lazily in the activity and access it in a `LaunchedEffect,` it's all good. But if we access it directly in `KoinAndroidContext` (same place where `koinViewModel()` is called), it crashes. Also, if we reduce the number of `RootViewModel` dependencies, it doesn't seem to crash anymore.
```kotlin
class MainActivity : ComponentActivity() {
    private val viewModel: RootViewModel by viewModel()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContent {
            KoinAndroidContext {
            	Timber.d("viewModel: %s", viewModel) // this crashes

            	LaunchedEffect(Unit) {
            	    Timber.d("viewModel: %s", viewModel) // this doesn't crash
            	}
            }
        }
    }
}
```

We haven't found any acceptable workaround, so we need to go back to 3.5.x for now.

**Comment by wandfliege** [comment]
@arnaudgiuliani 
Not sure if directly related. This error happened in the verifcation test for my modules. 

Version catalogue definitions
koinAnnotations = "2.0.0"
koin = "4.0.4"

Note I replaced my 
```
org.koin.test.verify.MissingKoinDefinitionException: Missing definition for '[field:'savedStateHandle' - type:'androidx.lifecycle.SavedStateHandle']' in definition '[Factory: 'com....*ViewModel']'.
	at app//org.koin.test.verify.Verification.verifyFactory(Verification.kt:82)
	at app//org.koin.test.verify.Verification.verify(Verification.kt:46)
	at app//org.koin.test.verify.Verify.verify(VerifyModule.kt:81)
	at app//org.koin.test.verify.VerifyModuleKt.verify(VerifyModule.kt:30)
	at app//org.koin.test.verify.VerifyModuleKt.verify$default(VerifyModule.kt:30)
	at app//com....*AppModuleTest.should verify all modules(AppModuleTest.kt:13)
	at java.base@21.0.6/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base@21.0.6/java.lang.reflect.Method.invoke(Method.java:580)
	at app//org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at app//org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at app//org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at app//org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at app//org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at app//org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at app//org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at app//org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at app//org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at app//org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at app//org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at app//org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at app//org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at app//org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at app//org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at app//org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:112)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:40)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:54)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:53)
	at java.base@21.0.6/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base@21.0.6/java.lang.reflect.Method.invoke(Method.java:580)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:92)
	at jdk.proxy1/jdk.proxy1.$Proxy4.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:183)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:132)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:103)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:63)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:121)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

ViewModel
```
@KoinViewModel
internal class ...*ViewModel(
    savedStateHandle: SavedStateHandle,
) { ... }
```

Generated code
```
@ExternalDefinition("...")
public fun Module.define...*ViewModel() : KoinDefinition<*> = viewModel() { _ -> ...*ViewModel(savedStateHandle=get()) } 

public val _defaultModule : Module get() = module {
	define...*ViewModel()
}
public val defaultModule : org.koin.core.module.Module get() = _defaultModule
public fun org.koin.core.KoinApplication.defaultModule(): org.koin.core.KoinApplication = modules(defaultModule)
```

**Comment by arnaudgiuliani** [comment]
> Kind reminder that this issue still exists in version `4.0.4`.
> 
> Unfortunately, it occurs randomly and only in release builds, so I haven’t been able to reproduce it reliably.
> 
> The error seems to be triggered here, which suggests that `ResolutionContext` might be broken for `SavedStateHandle`:
> 
> ```
> private fun <T> resolveFromContext(
>      instanceContext: ResolutionContext
> ): T {
>      return resolveFromInjectedParameters(instanceContext)
>          ?: resolveFromRegistry(instanceContext)
>          ?: resolveFromStackedParameters(instanceContext)
>          ?: resolveFromScopeSource(instanceContext)
>          ?: resolveFromParentScopes(instanceContext)
>          ?: throwNoDefinitionFound(instanceContext)
> }
> ```
> 
> Might be there some simple solution to add savedStateHandle explicitly `viewModel { savedStateHandle: SavedStateHandle -> }` or determine scope of `ViewModel` for `SavedStateHandle`?

Any library update in Android around? in 4.1.1 we have some fixes

**Comment by kzotin** [comment]
just FYI, we've got the same issue as @mars885 in 1M+ DAU app, and downgraded to `3.5.6`

**Comment by ilkersi** [comment]
This is effecting our app as well, like @mars885 and @kzotin stated. We downgraded to `3.5.6` and the issue disappeared. Can you please re-open this issue @arnaudgiuliani?

**Comment by arnaudgiuliani** [comment]
Can any of you create a sample app to help reproduce the issue?

**Comment by ilkersi** [comment]
It is difficult to replicate with a sample app, but I can provide you the stacktrace of the crash that happened yesterday:

```
at android.app.ActivityThread.performLaunchActivityActivityThread.java:4458
at android.app.ActivityThread.handleLaunchActivityActivityThread.java:4689
at android.app.servertransaction.LaunchActivityItem.executeLaunchActivityItem.java:112
at android.app.servertransaction.TransactionExecutor.executeNonLifecycleItemTransactionExecutor.java:174
at android.app.servertransaction.TransactionExecutor.executeTransactionItemsTransactionExecutor.java:109
at android.app.servertransaction.TransactionExecutor.executeTransactionExecutor.java:81
at android.app.ActivityThread$H.handleMessageActivityThread.java:2899
at android.os.Handler.dispatchMessageHandler.java:107
at android.os.Looper.loopOnceLooper.java:257
at android.os.Looper.loopLooper.java:342
at android.app.ActivityThread.mainActivityThread.java:9638
at java.lang.reflect.Method.invokeNative Method
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.runRuntimeInit.java:619
at com.android.internal.os.ZygoteInit.mainZygoteInit.java:929
Caused by: org.koin.core.error.InstanceCreationException: Could not create instance for '[Factory: 'MainActivityViewModel']'
at org.koin.core.instance.InstanceFactory.createInstanceFactory.kt:56
at org.koin.core.instance.FactoryInstanceFactory.getFactoryInstanceFactory.kt:38
at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_coreInstanceRegistry.kt:132
at org.koin.core.resolution.CoreResolver.resolveFromRegistryCoreResolver.kt:87
at org.koin.core.resolution.CoreResolver.resolveFromContextOrNullCoreResolver.kt:74
at org.koin.core.resolution.CoreResolver.resolveFromContextOrNull$defaultCoreResolver.kt:72
at org.koin.core.resolution.CoreResolver.resolveFromContextCoreResolver.kt:69
at org.koin.core.scope.Scope.resolveFromContextScope.kt:321
at org.koin.core.scope.Scope.stackParametersCallScope.kt:292
at org.koin.core.scope.Scope.resolveInstanceScope.kt:270
at org.koin.core.scope.Scope.resolveScope.kt:243
at org.koin.core.scope.Scope.getWithParametersScope.kt:234
at org.koin.viewmodel.factory.KoinViewModelFactory.createKoinViewModelFactory.kt:49
at androidx.lifecycle.viewmodel.ViewModelProviderImpl_androidKt.createViewModelViewModelProviderImpl.android.kt:34
at androidx.lifecycle.viewmodel.ViewModelProviderImpl.getViewModel$lifecycle_viewmodel_releaseViewModelProviderImpl.kt:60
at androidx.lifecycle.viewmodel.ViewModelProviderImpl.getViewModel$lifecycle_viewmodel_release$defaultViewModelProviderImpl.kt:43
at androidx.lifecycle.ViewModelProvider.getViewModelProvider.android.kt:92
at org.koin.viewmodel.GetViewModelKt.resolveViewModelGetViewModel.kt:60
at org.koin.viewmodel.GetViewModelKt.resolveViewModel$defaultGetViewModel.kt:44
at MainActivity.onCreateMainActivity.kt:325
at android.app.Activity.performCreateActivity.java:9363
at android.app.Activity.performCreateActivity.java:9332
at android.app.Instrumentation.callActivityOnCreateInstrumentation.java:1526
at android.app.ActivityThread.performLaunchActivityActivityThread.java:4440
... 13 more
**Caused by: org.koin.core.error.NoDefinitionFoundException: No definition found for type 'androidx.lifecycle.SavedStateHandle'. Check your Modules configuration and add missing type and/or qualifier!**
at org.koin.core.resolution.CoreResolver.resolveFromContextCoreResolver.kt:178
at org.koin.core.scope.Scope.resolveFromContextScope.kt:321
at org.koin.core.scope.Scope.stackParametersCallScope.kt:284
at org.koin.core.scope.Scope.resolveInstanceScope.kt:270
at org.koin.core.scope.Scope.resolveScope.kt:243
at org.koin.core.scope.Scope.getScope.kt:225
at koinAppModule$lambda$16$lambda$13KoinAppModule.kt:286	
at KoinAppModuleKt.$r8$lambda$cBrGSUvD3OmkFrv6ybfoepi8iA0(Unknown Source:0)
at KoinAppModuleKt$$ExternalSyntheticLambda15.invokeD8$$SyntheticClass
at org.koin.core.instance.InstanceFactory.createInstanceFactory.kt:49
... 36 more
```

**Comment by arnaudgiuliani** [comment]
Do you use parameters injected with the ViewModel? There is eitehr something interfering, either something else ... but SavedState Injection is working since a while. That's really strange 🤔

**Comment by ilkersi** [comment]
Yes, the `MainActivityViewModel` in this case has two injected parameters which we can only get after the viewmodel initialization. And also there is another constructor parameter which is `Lazy<SomeInterface>`.

So to sum up it looks like this:
```
viewModel { parameters ->
    MainActivityViewModel(
        a = parameters.get(),
        b = parameters.get(),
        c = lazy { get() }
    )
}

class MainActivityViewModel(
    private val a: A,
    private val b: B,
    private val c: Lazy<C>
)
```

**Comment by andreiJJFS** [comment]
Also getting this issue in production, but only some of the time. Haven't been able to reproduce it but can see the error in Sentry.

`viewModelOf(::TemperatureViewModel)`

`@Composable
fun TemperatureScreen(viewModel: TemperatureViewModel = koinViewModel()`

`class TemperatureViewModel(handle: SavedStateHandle): ViewModel() {...}`

<details>
<summary>Click to expand stack trace</summary>

 org.koin.core.error.NoDefinitionFoundException: No definition found for type 'androidx.lifecycle.SavedStateHandle'. Check your Modules configuration and add missing type and/or qualifier!
    at org.koin.core.scope.Scope.resolveFromContext(Scope.kt:535)
    at org.koin.core.scope.Scope.stackParametersCall(Scope.kt:274)
    at org.koin.core.scope.Scope.resolveInstance(Scope.kt:260)
    at org.koin.core.scope.Scope.resolve(Scope.kt:233)
    at org.koin.core.scope.Scope.get(Scope.kt:215)
    at 
di.AppConfig$viewModelModules$lambda$17$$inlined$viewModelOf$default$5.invoke(ViewModelOf.kt:254)
    at 
.di.AppConfig$viewModelModules$lambda$17$$inlined$viewModelOf$default$5.invoke(ViewModelOf.kt:94)
    at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:51)
    at org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:38)
    at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:111)
    at org.koin.core.scope.Scope.resolveFromRegistry(Scope.kt:322)
    at org.koin.core.scope.Scope.resolveFromContext(Scope.kt:312)
    at org.koin.core.scope.Scope.stackParametersCall(Scope.kt:282)
    at org.koin.core.scope.Scope.resolveInstance(Scope.kt:260)
    at org.koin.core.scope.Scope.resolve(Scope.kt:233)
    at org.koin.core.scope.Scope.getWithParameters(Scope.kt:224)
    at org.koin.viewmodel.factory.KoinViewModelFactory.create(KoinViewModelFactory.kt:41)
    at androidx.lifecycle.viewmodel.ViewModelProviderImpl_androidKt.createViewModel(ViewModelProviderImpl.android.kt:34)
    at androidx.lifecycle.viewmodel.ViewModelProviderImpl.getViewModel$lifecycle_viewmodel_release(ViewModelProviderImpl.kt:65)
    at androidx.lifecycle.viewmodel.ViewModelProviderImpl.getViewModel$lifecycle_viewmodel_release$default(ViewModelProviderImpl.kt:47)
    at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.android.kt:91)
    at org.koin.viewmodel.GetViewModelKt.resolveViewModel(GetViewModel.kt:60)
    at .ui.screens.driving.DrivingScreenKt.DrivingScreen(DrivingScreen.kt:280)
    at .ui.navigation.NavigationKt$mainGraph$1$2.invoke(Navigation.kt:285)
    at .ui.navigation.NavigationKt$mainGraph$1$2.invoke(Navigation.kt:284)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:139)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.navigation.compose.NavHostKt$NavHost$32$1.invoke(NavHost.kt:704)
    at androidx.navigation.compose.NavHostKt$NavHost$32$1.invoke(NavHost.kt:703)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:109)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.compose.runtime.CompositionLocalKt.CompositionLocalProvider(CompositionLocal.kt:401)
    at androidx.compose.runtime.saveable.SaveableStateHolderImpl.SaveableStateProvider(SaveableStateHolder.kt:85)
    at androidx.navigation.compose.NavBackStackEntryProviderKt.SaveableStateProvider(NavBackStackEntryProvider.kt:65)
    at androidx.navigation.compose.NavBackStackEntryProviderKt.access$SaveableStateProvider(NavBackStackEntryProvider.kt:1)
    at androidx.navigation.compose.NavBackStackEntryProviderKt$LocalOwnersProvider$1.invoke(NavBackStackEntryProvider.kt:52)
    at androidx.navigation.compose.NavBackStackEntryProviderKt$LocalOwnersProvider$1.invoke(NavBackStackEntryProvider.kt:51)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:109)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.compose.runtime.CompositionLocalKt.CompositionLocalProvider(CompositionLocal.kt:380)
    at androidx.navigation.compose.NavBackStackEntryProviderKt.LocalOwnersProvider(NavBackStackEntryProvider.kt:47)
    at androidx.navigation.compose.NavHostKt$NavHost$32.invoke(NavHost.kt:703)
    at androidx.navigation.compose.NavHostKt$NavHost$32.invoke(NavHost.kt:683)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:139)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1$5.invoke(AnimatedContent.kt:803)
    at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1$5.invoke(AnimatedContent.kt:792)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:118)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.compose.animation.AnimatedVisibilityKt.AnimatedEnterExitImpl(AnimatedVisibility.kt:771)
    at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1.invoke(AnimatedContent.kt:774)
    at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1.invoke(AnimatedContent.kt:757)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:109)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.compose.animation.AnimatedContentKt.AnimatedContent(AnimatedContent.kt:816)
    at androidx.navigation.compose.NavHostKt.NavHost(NavHost.kt:655)
    at androidx.navigation.compose.NavHostKt$NavHost$34.invoke(:29)
    at androidx.navigation.compose.NavHostKt$NavHost$34.invoke(:8)
    at androidx.compose.runtime.RecomposeScopeImpl.compose(RecomposeScopeImpl.kt:192)
    at androidx.compose.runtime.ComposerImpl.recomposeToGroupEnd(Composer.kt:2825)
    at androidx.compose.runtime.ComposerImpl.skipCurrentGroup(Composer.kt:3116)
    at androidx.compose.runtime.ComposerImpl.doCompose(Composer.kt:3607)
    at androidx.compose.runtime.ComposerImpl.recompose$runtime_release(Composer.kt:3552)
    at androidx.compose.runtime.CompositionImpl.recompose(Composition.kt:948)
    at androidx.compose.runtime.Recomposer.performRecompose(Recomposer.kt:1206)
    at androidx.compose.runtime.Recomposer.access$performRecompose(Recomposer.kt:132)
    at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:616)
    at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:585)
    at androidx.compose.ui.platform.AndroidUiFrameClock$withFrameNanos$2$callback$1.doFrame(AndroidUiFrameClock.android.kt:41)
    at androidx.compose.ui.platform.AndroidUiDispatcher.performFrameDispatch(AndroidUiDispatcher.android.kt:109)
    at androidx.compose.ui.platform.AndroidUiDispatcher.access$performFrameDispatch(AndroidUiDispatcher.android.kt:41)
    at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.doFrame(AndroidUiDispatcher.android.kt:69)
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1397)
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1408)
    at android.view.Choreographer.doCallbacks(Choreographer.java:1012)
    at android.view.Choreographer.doFrame(Choreographer.java:938)
    at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1382)
    at android.os.Handler.handleCallback(Handler.java:958)
    at android.os.Handler.dispatchMessage(Handler.java:99)
    at android.os.Looper.loopOnce(Looper.java:205)
    at android.os.Looper.loop(Looper.java:294)
    at android.app.ActivityThread.main(ActivityThread.java:8385)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:640)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:982)

org.koin.core.error.InstanceCreationException: Could not create instance for '[Factory: '.ui.screens.driving.DrivingViewModel']'
    at org.koin.core.instance.InstanceFactory.create(InstanceFactory.kt:58)
    at org.koin.core.instance.FactoryInstanceFactory.get(FactoryInstanceFactory.kt:38)
    at org.koin.core.registry.InstanceRegistry.resolveInstance$koin_core(InstanceRegistry.kt:111)
    at org.koin.core.scope.Scope.resolveFromRegistry(Scope.kt:322)
    at org.koin.core.scope.Scope.resolveFromContext(Scope.kt:312)
    at org.koin.core.scope.Scope.stackParametersCall(Scope.kt:282)
    at org.koin.core.scope.Scope.resolveInstance(Scope.kt:260)
    at org.koin.core.scope.Scope.resolve(Scope.kt:233)
    at org.koin.core.scope.Scope.getWithParameters(Scope.kt:224)
    at org.koin.viewmodel.factory.KoinViewModelFactory.create(KoinViewModelFactory.kt:41)
    at androidx.lifecycle.viewmodel.ViewModelProviderImpl_androidKt.createViewModel(ViewModelProviderImpl.android.kt:34)
    at androidx.lifecycle.viewmodel.ViewModelProviderImpl.getViewModel$lifecycle_viewmodel_release(ViewModelProviderImpl.kt:65)
    at androidx.lifecycle.viewmodel.ViewModelProviderImpl.getViewModel$lifecycle_viewmodel_release$default(ViewModelProviderImpl.kt:47)
    at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.android.kt:91)
    at org.koin.viewmodel.GetViewModelKt.resolveViewModel(GetViewModel.kt:60)
    at .ui.screens.driving.DrivingScreenKt.DrivingScreen(DrivingScreen.kt:280)
    at .ui.navigation.NavigationKt$mainGraph$1$2.invoke(Navigation.kt:285)
    at .ui.navigation.NavigationKt$mainGraph$1$2.invoke(Navigation.kt:284)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:139)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.navigation.compose.NavHostKt$NavHost$32$1.invoke(NavHost.kt:704)
    at androidx.navigation.compose.NavHostKt$NavHost$32$1.invoke(NavHost.kt:703)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:109)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.compose.runtime.CompositionLocalKt.CompositionLocalProvider(CompositionLocal.kt:401)
    at androidx.compose.runtime.saveable.SaveableStateHolderImpl.SaveableStateProvider(SaveableStateHolder.kt:85)
    at androidx.navigation.compose.NavBackStackEntryProviderKt.SaveableStateProvider(NavBackStackEntryProvider.kt:65)
    at androidx.navigation.compose.NavBackStackEntryProviderKt.access$SaveableStateProvider(NavBackStackEntryProvider.kt:1)
    at androidx.navigation.compose.NavBackStackEntryProviderKt$LocalOwnersProvider$1.invoke(NavBackStackEntryProvider.kt:52)
    at androidx.navigation.compose.NavBackStackEntryProviderKt$LocalOwnersProvider$1.invoke(NavBackStackEntryProvider.kt:51)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:109)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.compose.runtime.CompositionLocalKt.CompositionLocalProvider(CompositionLocal.kt:380)
    at androidx.navigation.compose.NavBackStackEntryProviderKt.LocalOwnersProvider(NavBackStackEntryProvider.kt:47)
    at androidx.navigation.compose.NavHostKt$NavHost$32.invoke(NavHost.kt:703)
    at androidx.navigation.compose.NavHostKt$NavHost$32.invoke(NavHost.kt:683)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:139)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1$5.invoke(AnimatedContent.kt:803)
    at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1$5.invoke(AnimatedContent.kt:792)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:118)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.compose.animation.AnimatedVisibilityKt.AnimatedEnterExitImpl(AnimatedVisibility.kt:771)
    at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1.invoke(AnimatedContent.kt:774)
    at androidx.compose.animation.AnimatedContentKt$AnimatedContent$6$1.invoke(AnimatedContent.kt:757)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:109)
    at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.jvm.kt:35)
    at androidx.compose.animation.AnimatedContentKt.AnimatedContent(AnimatedContent.kt:816)
    at androidx.navigation.compose.NavHostKt.NavHost(NavHost.kt:655)
    at androidx.navigation.compose.NavHostKt$NavHost$34.invoke(:29)
    at androidx.navigation.compose.NavHostKt$NavHost$34.invoke(:8)
    at androidx.compose.runtime.RecomposeScopeImpl.compose(RecomposeScopeImpl.kt:192)
    at androidx.compose.runtime.ComposerImpl.recomposeToGroupEnd(Composer.kt:2825)
    at androidx.compose.runtime.ComposerImpl.skipCurrentGroup(Composer.kt:3116)
    at androidx.compose.runtime.ComposerImpl.doCompose(Composer.kt:3607)
    at androidx.compose.runtime.ComposerImpl.recompose$runtime_release(Composer.kt:3552)
    at androidx.compose.runtime.CompositionImpl.recompose(Composition.kt:948)
    at androidx.compose.runtime.Recomposer.performRecompose(Recomposer.kt:1206)
    at androidx.compose.runtime.Recomposer.access$performRecompose(Recomposer.kt:132)
    at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:616)
    at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:585)
    at androidx.compose.ui.platform.AndroidUiFrameClock$withFrameNanos$2$callback$1.doFrame(AndroidUiFrameClock.android.kt:41)
    at androidx.compose.ui.platform.AndroidUiDispatcher.performFrameDispatch(AndroidUiDispatcher.android.kt:109)
    at androidx.compose.ui.platform.AndroidUiDispatcher.access$performFrameDispatch(AndroidUiDispatcher.android.kt:41)
    at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.doFrame(AndroidUiDispatcher.android.kt:69)
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1397)
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1408)
    at android.view.Choreographer.doCallbacks(Choreographer.java:1012)
    at android.view.Choreographer.doFrame(Choreographer.java:938)
    at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1382)
    at android.os.Handler.handleCallback(Handler.java:958)
    at android.os.Handler.dispatchMessage(Handler.java:99)
    at android.os.Looper.loopOnce(Looper.java:205)
    at android.os.Looper.loop(Looper.java:294)
    at android.app.ActivityThread.main(ActivityThread.java:8385)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:640)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:982)
</details>

**Comment by arnaudgiuliani** [comment]
We might improve SSH parameter detection with Compiler Plugin then.

**Comment by DarioAhdoot** [comment]
Just want to add that we are also seeing this in production. It's not prevalent and only occurs a small percentage of the time. However, given our volume of our users, the total number of crashes is high. We recently migrated from Dagger/Hilt to Koin, and this is our first deploy using Koin.

**Comment by ilkersi** [comment]
Any updates on this? We are blocked on the version `3.5.6` due to this bug.

**Comment by arnaudgiuliani** [comment]
I need ways to reproduce it. It's not something that can be spotted in average/current apps. Hard for me to help you here :/

**Comment by AbolfaZlRezaEe** [comment]
I was hoping that when I reach the bottom of this issue, I can see the fixed version or a workaround. But life isn't fair :(

downgrading...

**Comment by arnaudgiuliani** [comment]
minification is part of that

**Comment by ilkersi** [comment]
Any updates on this? I mean it's almost been 2 years.

**Comment by kzotin** [comment]
claude-fable might be capable enough

**Comment by arnaudgiuliani** [comment]
Update for **4.2.2**: we improved the diagnostics and shipped R8 guidance around this, but want to be transparent that the intermittent crash itself isn't a single fixable bug on Koin's side:

- `get<SavedStateHandle>()` now gives an **actionable error** when the ViewModel's `CreationExtras` has no `SavedStateRegistryOwner`, and the `NoDefinitionFound` message for `SavedStateHandle` now explains it comes from CreationExtras (not a module definition).
- New **[R8/ProGuard guide](https://insert-koin.io/docs/reference/koin-android/r8-proguard)** — importantly, Koin's ViewModel resolution is **R8-safe** (compile-time, no constructor reflection), so 'keep your ViewModels' rules are *not* the fix. What you do keep: your own saved-state (`@Parcelize`/`Serializable`) classes for process-death restore.
- `SavedStateHandle` must be injected in the **constructor** — it's only available while the ViewModel is created (deferred/`lazy { get() }` can't work and now errors clearly).

Keeping this open: the intermittent process-death reports aren't deterministically reproducible (the extras lifecycle is androidx-side). If you can share a minimal repro that reproduces on **4.2.2**, please attach it. Otherwise the above should turn the mystery crash into a clear, fixable setup error.
